### DISCUSSION TITLE
How to use labgrid-client to flash image on remote pc?

### DISCUSSION THREAD

**Discussion by daocloud123** [labgrid-project/labgrid#951] [General]
When we use `labgrid-client -p xxx fastboot` or something similar, how could it works if we execute `labgrid-client` on a machine which not directly link the board?

If possible you could give me an example configure of `exporter` for such resource? And suggest me the full `labgrid-client command` example? Thanks!

Additional, does `fastboot` runs on the machine which execute `labgrid-client` or runs on the remote resource?

**Comment by Emantor** [comment]
> When we use `labgrid-client -p xxx fastboot` or something similar, how could it works if we execute `labgrid-client` on a machine which not directly link the board?

labgrid copies the file to the machine where the board is connected to (the exporter) and than does an SSH onto the remote machine and executes fastboot with the copied file.

> If possible you could give me an example configure of `exporter` for such resource? And suggest me the full `labgrid-client command` example? Thanks!
```
exporter-example-group:
  location: "In an example location"
  AndroidFastboot:
    match:
      '@ID_PATH': 'pci-000:07:00.0-usb-0:1.3:1.0'
```
The udev path above was taken from one of my exporters. If you have a fastboot device connected to the exporter, you can simply run `labgrid-suggest` to get the correct entry for the exporter YAML.

Additionally you'll want to look into the introductory videos from my colleague over on the pengutronix [youtube channel](https://www.youtube.com/watch?v=In_mOO35V7c&list=PLPy6BX4pJosCEq7CIU06bt2WM7lFAy1CF).

> 
> Additional, does `fastboot` runs on the machine which execute `labgrid-client` or runs on the remote resource?

It runs on the remote machine.

**Reply by daocloud123** [reply]
This is really useful! A final question before I go to the youtube:

If possible I could configure the infrastructure as next?

1. machineA (directly link one device to this machine)
2. machineB (runs exporter on this machine)
3. machineC (runs labgrid-client)

Then on `machineC`, I run  the `labgrid-client`, then labgrid somehow get the configure from `machineB`, and send images to `machineA`, finally runs `fastboot` on `machineA`?

In a word, the exporter not runs on the machine which directly link the board, is it possible?

More, if `fastboot` not support that, do we have other resource support such configure? Or this is not as design, and should be avoid for all kinds of resources?

**Reply by Bastian-Krause** [reply]
> If possible I could configure the infrastructure as next?
> 
> 1. machineA (directly link one device to this machine)
> 2. machineB (runs exporter on this machine)
> 3. machineC (runs labgrid-client)
> 
> Then on `machineC`, I run the `labgrid-client`, then labgrid somehow get the configure from `machineB`, and send images to `machineA`, finally runs `fastboot` on `machineA`?
> 
> In a word, the exporter not runs on the machine which directly link the board, is it possible?

No. The exporter exports (hence the name) its resources, so there is no point in running it on a different machine.

> More, if `fastboot` not support that, do we have other resource support such configure? Or this is not as design, and should be avoid for all kinds of resources?

This applies to all resources, except for stand-alone network resources (i.e. resources with a `host` attribute). Have a look at [labgrid's Resources Documentation](https://labgrid.readthedocs.io/en/latest/configuration.html#resources) for a list of all available resource types.
Also see the [Architecture Overview](https://labgrid.readthedocs.io/en/latest/overview.html) to better understand what roles the different components play and how they interact.

**Reply by daocloud123** [reply]
Thank you, this answered my concern!
